### ISSUE TITLE
ThreadSanitizer data race warning

### ISSUE THREAD

**Issue by tiennou** [libgit2/libgit2#4313]
Got that while running the test suite with ThreadSanitizer enabled, on current `master`.

```
object::cache..==================
WARNING: ThreadSanitizer: data race (pid=68977)
  Read of size 4 at 0x7d4c00008a18 by thread T4:
    #0 git_repository__cvar config_cache.c:107 (libgit2_clar+0x000100054293)
    #1 git_odb__set_caps odb.c:637 (libgit2_clar+0x0001000f1ed1)
    #2 git_repository_odb__weakptr repository.c:1084 (libgit2_clar+0x00010016387e)
    #3 git_object_lookup_prefix object.c:129 (libgit2_clar+0x0001000ede1f)
    #4 git_object_lookup object.c:198 (libgit2_clar+0x0001000ee26c)
    #5 cache_parsed cache.c:151 (libgit2_clar+0x000100324435)

  Previous write of size 4 at 0x7d4c00008a18 by thread T1:
    #0 git_repository__cvar config_cache.c:117 (libgit2_clar+0x000100054380)
    #1 git_odb__set_caps odb.c:637 (libgit2_clar+0x0001000f1ed1)
    #2 git_repository_odb__weakptr repository.c:1084 (libgit2_clar+0x00010016387e)
    #3 git_repository_odb repository.c:1105 (libgit2_clar+0x000100163c98)
    #4 cache_raw cache.c:173 (libgit2_clar+0x000100324754)

  Location is heap block of size 424 at 0x7d4c00008880 allocated by main thread:
    #0 calloc <null>:144 (libclang_rt.tsan_osx_dynamic.dylib+0x000000041a42)
    #1 git__calloc util.h:169 (libgit2_clar+0x000100169dd8)
    #2 repository_alloc repository.c:229 (libgit2_clar+0x00010015ff03)
    #3 git_repository_open_ext repository.c:811 (libgit2_clar+0x0001001607a4)
    #4 git_repository_open repository.c:867 (libgit2_clar+0x0001001628bf)
    #5 test_object_cache__threadmania cache.c:212 (libgit2_clar+0x000100323f3d)
    #6 clar_run_test clar.c:222 (libgit2_clar+0x00010021ae0f)
    #7 clar_run_suite clar.c:286 (libgit2_clar+0x000100218410)
    #8 clar_test_run clar.c:426 (libgit2_clar+0x0001002178d7)
    #9 main main.c:19 (libgit2_clar+0x0001002efc52)

  Thread T4 (tid=1953212, running) created by main thread at:
    #0 pthread_create <null>:144 (libclang_rt.tsan_osx_dynamic.dylib+0x000000024410)
    #1 test_object_cache__threadmania cache.c:223 (libgit2_clar+0x0001003240d1)
    #2 clar_run_test clar.c:222 (libgit2_clar+0x00010021ae0f)
    #3 clar_run_suite clar.c:286 (libgit2_clar+0x000100218410)
    #4 clar_test_run clar.c:426 (libgit2_clar+0x0001002178d7)
    #5 main main.c:19 (libgit2_clar+0x0001002efc52)

  Thread T1 (tid=1953211, running) created by main thread at:
    #0 pthread_create <null>:144 (libclang_rt.tsan_osx_dynamic.dylib+0x000000024410)
    #1 test_object_cache__threadmania cache.c:223 (libgit2_clar+0x0001003240d1)
    #2 clar_run_test clar.c:222 (libgit2_clar+0x00010021ae0f)
    #3 clar_run_suite clar.c:286 (libgit2_clar+0x000100218410)
    #4 clar_test_run clar.c:426 (libgit2_clar+0x0001002178d7)
    #5 main main.c:19 (libgit2_clar+0x0001002efc52)

SUMMARY: ThreadSanitizer: data race config_cache.c:107 in git_repository__cvar
==================
```

**Comment by tiennou** [comment]
Complete report : https://gist.github.com/tiennou/07a900711c2ef3d855561dd59b30e973

**Comment by pks-t** [comment]
Yeah, threading... thing is, our threading statement (THREADING.md) explicitly says:

```
Use an object from a single thread at a time. Most data structures do
not guard against concurrent access themselves. This is because they
are rarely used in isolation and it makes more sense to synchronize
access via a larger lock or similar mechanism.
```

So our concurrent threadmania and diff tests simply disregard our own statement completly and test this regardless. I'm always being optimistic here and think we might improve upon the situation later on, which would lead us to making more favorable guarantees. But until now, whenever I tried to fix races I just ran into a plethora of other races.

TL;DR: we do not support accessing shared objects at the same time from different threads. We should delete those tests. :(

**Comment by tiennou** [comment]
Should we instead fix the tests so they respect the statement then ?

**Comment by pks-t** [comment]
Where "fix" can only mean "delete", unfotunately. But I guess so. @ethomson?

**Comment by tiennou** [comment]
Well, I meant, create new separate git_repositories each. At least that keeps the statement true, and tests something which should be guaranteed.

(Relatedly, there's a reference test that deletes `HEAD`, which has pretty strange semantics too)

**Comment by ethomson** [comment]
We’ve updated these tests to reflect our threading policies.
